File: app/src/main/java/life/plank/juna/zone/ZoneApplication.java
package life.plank.juna.zone;

import android.app.Application;
import android.content.Context;

import com.crashlytics.android.Crashlytics;
import com.google.gson.Gson;

import io.alterac.blurkit.BlurKit;
import io.fabric.sdk.android.Fabric;
import life.plank.juna.zone.injection.component.AppComponent;
import life.plank.juna.zone.injection.component.DaggerAppComponent;
import life.plank.juna.zone.injection.component.DaggerViewModelComponent;
import life.plank.juna.zone.injection.component.NetworkComponent;
import life.plank.juna.zone.injection.component.UiComponent;
import life.plank.juna.zone.injection.component.ViewModelComponent;
import life.plank.juna.zone.injection.module.ContextModule;

public class ZoneApplication extends Application {

    private static ZoneApplication zoneApplication;

    private AppComponent appComponent;
    private UiComponent uiComponent;
    private NetworkComponent networkComponent;
    private ViewModelComponent viewModelComponent;

    public static ZoneApplication getApplication() {
        return zoneApplication;
    }

    public static Context getContext() {
        return getApplication().getApplicationContext();
    }

    @Override public void onCreate() {
        super.onCreate();
        zoneApplication = this;
        Fabric.with(this, new Crashlytics());
        BlurKit.init(this);

        appComponent = DaggerAppComponent.builder()
                .contextModule(new ContextModule(getApplicationContext()))
                .build();
        networkComponent = appComponent.networkComponentBuilder()
                .build();
        uiComponent = networkComponent.viewComponentBuilder()
                .build();
        viewModelComponent = DaggerViewModelComponent.builder().build();
    }

    public static Gson getGson() {
        return zoneApplication.networkComponent.getGson();
    }

    public AppComponent getAppComponent() {
        return appComponent;
    }

    public NetworkComponent getNetworkComponent() {
        return networkComponent;
    }

    public UiComponent getUiComponent() {
        return uiComponent;
    }

    public ViewModelComponent getViewModelComponent() {
        return viewModelComponent;
    }
}

File: app/src/main/java/life/plank/juna/zone/ZoneApplication.kt
package life.plank.juna.zone

import android.app.Application
import android.content.Context
import com.crashlytics.android.Crashlytics
import com.google.gson.Gson
import io.alterac.blurkit.BlurKit
import io.fabric.sdk.android.Fabric
import life.plank.juna.zone.injection.component.*
import life.plank.juna.zone.injection.module.ContextModule

class ZoneApplication : Application() {

    lateinit var appComponent: AppComponent
        private set
    lateinit var uiComponent: UiComponent
        private set
    lateinit var networkComponent: NetworkComponent
        private set
    lateinit var viewModelComponent: ViewModelComponent
        private set

    companion object {
        lateinit var application: ZoneApplication
            private set

        val appContext: Context
            get() = application.applicationContext

        val gson: Gson
            get() = application.networkComponent.gson
    }

    override fun onCreate() {
        super.onCreate()
        application = this
        Fabric.with(this, Crashlytics())
        BlurKit.init(this)

        appComponent = DaggerAppComponent.builder()
                .contextModule(ContextModule(applicationContext))
                .build()
        networkComponent = appComponent.networkComponentBuilder()
                .build()
        uiComponent = networkComponent.viewComponentBuilder()
                .build()
        viewModelComponent = DaggerViewModelComponent.builder().build()
    }
}